Rover: Fix SERVOx_REVERSED does not work as expected for ESC type "BrushedWithRelay"

tig987 opened the pull request:
See bug report #13269

ESCs of type PWM_TYPE_BRUSHED_WITH_RELAY need to set trim to min to set duty cycle range (0 - 100%) to servo range.
The existing function "set_trim_to_min_for" does set trim to max instead when the reversed flag is set, but for ESCs of type PWM_TYPE_BRUSHED_WITH_RELAY only the relay output should be inverted when reversed is set and not the PWM range.
Setting trim to max for PWM_TYPE_BRUSHED_WITH_RELAY caused the motor to run at full speed with no throttle.

Since the existing function "set_trim_to_min_for" is also used by the Ardupilot plane code I did not modify it but introduced a new function "set_trim_to_min2_for" which is used in my modification of the Ardupilot rover code now.

IamPete1 commented:
@tig987 this needs a rebase and seems like a reasonable feature, not a fan of `set_trim_to_min2_for`  function name, maybe something more verbose.

tig987 commented:
Hi,
I don't have much experience with github (this is my first contribution to a project with a pull request) but I will see that I get a rebase done.
I am also didn't like `set_trim_to_min2_for` much as a name, but at that point couldn't come up with something better.
Do you have a suggestion how to name the function?

IamPete1 commented:
@tig987 Thanks. I'm just having a look at the functions we already have, maybe you could have a new function that takes the limit enum. Something like:

```
    // set servo trim to a Limit
    static void set_trim_limit(SRV_Channel::Aux_servo_function_t function, SRV_Channel::Limit limit);
```

tig987 commented:
I took a look at the code and the limit enum `SRV_Channel::Limit limit` has the comment
`// used to get min/max/trim limit value based on reverse`
added to it.
So I guess that other developers would expect this new function to behave the same way and would expect Min/Max to be flipped on reversed.
For this purpose I added the parameter  `ignore_reversed ` to the function. It looks like this now:
`static void set_trim_to_limit_for(SRV_Channel::Aux_servo_function_t function, SRV_Channel::Limit limit, bool ignore_reversed = false);`
I also adjusted the naming to the other "set_trim" functions.

I completed the rebase - please have a look if I did it right.

IamPete1 commented:
Your best bet to fix this is to cherry pick the commits you want over to a new branch and force push over this one. It probably went wrong because of the merge commits we don't use them in AP. See https://ardupilot.org/dev/docs/git-rebase.html

tig987 commented:
OK, it seems I finally managed to rebase correctly, so that the build tests on Semaphore do not fail anymore.

tig987 commented:
@IamPete1 I saw your reply too late.
Is my pull request ok now, since it passed the Semaphore test, even if it contains a merge commit?

IamPete1 commented:
@tig987 Not fixed yet, look at the diff, 3890 commits, happy to help fixing if you can say the commits you need.

tig987 commented:
@IamPete1 many thanks for your offer to help.
I followed the instructions from  https://ardupilot.org/dev/docs/git-rebase.html and also tried an interactive rebase https://ardupilot.org/dev/docs/git-interactive-rebase.html
The "Merge branch 'master'" is always added automatically when I do the rebase.
I guess the 3890 commits are necessary because my fork and code changes are 1,5 years old?

Nevertheless the relevant commits are:

1. SRV_Channel: new function that allows setting trim to min regardless …  : https://github.com/ArduPilot/ardupilot/pull/13386/commits/1a96b7707e4a57f29605ad92923a24d92fae4ec4
2. new function set_trim_to_limit_for …  : https://github.com/ArduPilot/ardupilot/pull/13386/commits/092a04744218c10cfc56ad497af959249e907689
 
I would appreciate if you add them manually.

IamPete1 commented:
Its now those two commits on top of master.

tig987 commented:
@IamPete1 Thank you!

IamPete1 commented:
@tig987 Thanks for your work on this, and tracking down the source of the issue. https://github.com/ArduPilot/ardupilot/pull/18494 has been merged with a slightly re-worked fix.
